relayer/chains/substrate/event_parser.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap/zapcore"
+)
+
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
+// ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
+func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
+	for i := 0; i < len(ibcEvents); i++ {
+		for eType, data := range ibcEvents[i] {
+			var info ibcMessageInfo
+			var eventType string
+
+			marshalled, _ := json.Marshal(data)
+
+			switch eType {
+
+			case "CreateClient", "UpgradeClient", "ClientMisbehaviour":
+				ce := clientInfo{}
+				json.Unmarshal(marshalled, &ce)
+				info = &ce
+				eventType = eType
+
+			case "UpdateClient":
+				ce := clientUpdateInfo{}
+				json.Unmarshal(marshalled, &ce)
+				info = &ce
+				eventType = eType
+
+			case "OpenInitConnection", "OpenTryConnection", "OpenAckConnection", "OpenConfirmConnection":
+				ce := connectionEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &connectionInfo{
+					Height:               height,
+					ConnID:               ce.connectionID,
+					ClientID:             ce.clientID,
+					CounterpartyConnID:   ce.counterpartyConnectionID,
+					CounterpartyClientID: ce.counterpartyClientID,
+				}
+				eventType = eType
+
+			case "OpenInitChannel", "OpenTryChannel", "OpenAckChannel", "OpenConfirmChannel", "CloseInitChannel", "CloseConfirmChannel":
+				ce := channelEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &channelInfo{
+					Height:                height,
+					PortID:                ce.portID,
+					ChannelID:             strconv.FormatUint(ce.channelID, 10),
+					CounterpartyPortID:    strconv.FormatUint(ce.counterpartyPortID, 10),
+					CounterpartyChannelID: ce.counterpartyChannelID,
+					ConnID:                ce.connectionID,
+					CounterpartyConnID:    ce.counterpartyChannelID,
+					// TODO: how to populate order
+					// Order: ,
+					// TODO: how to populate version
+					// Version: ,
+				}
+				eventType = eType
+
+			case "SendPacket", "ReceivePacket", "WriteAcknowledgement", "AcknowledgePacket", "TimeoutPacket", "TimeoutOnClosePacket":
+				ce := packetEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &packetInfo{
+					Height:        height,
+					Sequence:      ce.sequence,
+					SourcePort:    ce.sourcePort,
+					SourceChannel: strconv.FormatUint(ce.sourceCannel, 10),
+					DestPort:      ce.destinationPort,
+					DestChannel:   strconv.FormatUint(ce.destinationChannel, 10),
+					// TODO: how to populate channel order
+					// ChannelOrder: ,
+					Data:             ce.data,
+					TimeoutHeight:    ce.timeoutHeight,
+					TimeoutTimestamp: ce.timeoutTimestamp,
+					// TODO: how to populate Ack
```

Comment:
The `Ack` bytes come from the `WriteAcknowledgement` event, while the rest of the params come from the `ReceivePacket` event. We use `packetAccumulator` for combining the info from both events in the cosmos event parser.

Reply:
Does it mean
```go
info = &packetInfo{Height: height}
```

Reply:
In cosmos, both the `WriteAcknowledgement` and `ReceivePacket` event will be included in a single IBC message. The `packetInfo` should be constructed with the `Height` property, as you have shown, if info is nil. If info is not nil, it should be casted to `packetInfo`. The attributes should be parsed into the resulting `packetInfo`. This will "accumulate" the properties from multiple events into a single `packetInfo`. The case of the `WriteAcknowledgement`/`ReceivePacket` events in the same message is the only need for the accumulator, the rest of the message types do not need an accumulator.

https://github.com/cosmos/relayer/blob/91182c9c766bfe3f2825255170a094dfd7bad287/relayer/chains/cosmos/event_parser.go#L98

Reply:
In this context, where is `msg`? 
Given that the function signature is
```go
func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
```

Reply:
Done, Please check the implementation.

Reply:
@agouin can you confirm, please?